torch/csrc/jit/passes/replacement_of_old_operators.cpp

```diff
@@ -0,0 +1,96 @@
+#include <torch/csrc/jit/passes/replacement_of_old_operators.h>
+
+#include <c10/util/Exception.h>
+#include <caffe2/serialize/versions.h>
+#include <torch/csrc/jit/ir/irparser.h>
+#include <torch/csrc/jit/operator_upgraders/upgraders.h>
+#include <torch/csrc/jit/operator_upgraders/utils.h>
+#include <torch/csrc/jit/operator_upgraders/version_map.h>
+#include <torch/csrc/jit/runtime/graph_iterator.h>
+#include <limits>
+#include <string>
+#include <unordered_map>
+
+namespace torch {
+namespace jit {
+
+static std::unordered_map<std::string, std::shared_ptr<Graph>> upgraderCache;
+
+struct OldOpsReplacer {
+  OldOpsReplacer(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {}
+
+  void run() {
+    if (!graph_->get_op_version().has_value()) {
+      return;
+    }
+    auto current_version = graph_->get_op_version().value();
+    DepthFirstGraphNodeIterator graph_it(graph_);
+    Node* node = graph_it.next();
+    while (node) {
+      if (auto schema = node->maybeSchema()) {
+        auto schema_name = schema->name() +
+            (schema->overload_name() != "" ? "." + schema->overload_name()
+                                           : "");
+        // this implies there was a version bump because of this operator
+        auto version_entry = get_operator_version_map().find(schema_name);
+        if (version_entry != get_operator_version_map().end()) {
+          const auto& entry = version_entry->second;
+          auto upgrader_entry =
+              findUpgrader(version_entry->second, current_version);
+          if (!upgrader_entry.has_value()) {
+            if (!isOpSymbolCurrent(schema_name, current_version)) {
+              TORCH_INTERNAL_ASSERT(
+                  false, "Upgrader must be present for ", schema_name);
+            }
+            node = graph_it.next();
+            continue;
+          }
+          auto upgrader_entry_val = upgrader_entry.value();
+          auto upgrader_name = upgrader_entry_val.upgrader_name;
+          auto upgrader_graph_entry = dump_upgraders_map().find(upgrader_name);
+          TORCH_INTERNAL_ASSERT(
+              upgrader_graph_entry != dump_upgraders_map().end(),
+              "Corresponding upgrader graph for ",
+              upgrader_name,
+              " must exist");
+
+          auto upgrader_graph = std::make_shared<Graph>();
+          auto upgrader_graph_it = upgraderCache.find(upgrader_name);
+          if (upgrader_graph_it != upgraderCache.end()) {
+            upgrader_graph = upgrader_graph_it->second;
+          } else {
+            parseIR(upgrader_graph_entry->second, upgrader_graph.get());
+            upgraderCache[upgrader_name] = upgrader_graph;
```

Comment:
It would be cleaner if we devise a helper function to read/update upgraderCache in a single call instead of complicating core logic of this pass with cache manipulations. Additionally, we can even compile all upgraders as first step of this pass for future uses.

Reply:
Moved it into an utility function.